### PR TITLE
Code Review For Django App

### PR DESCRIPTION
**Pull request by shadowshot-x**
@unishubh @shivam04  @haritha1313 

Please Review our code for django shopping cart and suggest changes

### CODE REVIEW

File: cart/cart.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from decimal import Decimal
```
Comment (shivam04) [inline cart/cart.py:1]: is this file for CRUD operation?

File: cart/templates/cart/cart.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+{% for final in finals %}
+{% if final.quantity %}
+
```
Comment (shivam04) [inline cart/templates/cart/cart.html:4]: add css

File: cart/templates/cart/detail.html [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                            {{ totail_items }} item {{ totail_items|pluralize }}, Rs. {{ cart.get_total_price }}
+                        </a>
+                        {% else %}
+                        Your cart is empty.
+                    {% endif %}
+                {% endwith %}
+            </span>
+        </h2>
+            <table class="table table-striped table-hover">
+                <thead style="background-color: #5AC8FA">
```
Comment (shivam04) [inline cart/templates/cart/detail.html:26]: add this style in some css file

File: cart/templates/cart/detail.html [hunk 4]
```diff
@@ -0,0 +1,75 @@
+{% extends 'shop/base.html' %}
+{% load static %}
+{% block title %}
+    Your Shopping Cart
+{% endblock %}
+
+
+{% block content %}
+    <div class="container-fluid bg6">
+        <div class="row cart-view" style="margin-top: 6%; margin-left: 70px; margin-right: 70px;">
```
Comment (shivam04) [inline cart/templates/cart/detail.html:10]: change px to %

File: cart/templates/cart/detail.html [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                        <th>Price</th>
+                    </tr>
+                </thead>
+                <tbody>
+                {% for item in cart %}
+                    {% with product=item.product  %}
+                        <tr>
+                            <td>
+                                <a href="{{ product.get_absolute_url }}">
+                                    <img src="{{product.image}}" alt="..." style="height: 100px; width: 200px;">
```
Comment (shivam04) [inline cart/templates/cart/detail.html:42]: px to %

File: cart/urls.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from django.conf.urls import url
+from . import views
+
+app_name = 'cart'
+
+urlpatterns = [
+    url(r'^$', views.cart_detail, name='cart_detail'),
+    url(r'^add/(?P<product_id>\d+)/$', views.cart_add, name='cart_add'),
```
Comment (shivam04) [inline cart/urls.py:8]: instead of id use slug

File: cart/views.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@require_POST
+def cart_add(request, product_id):
+    cart = Cart(request)  # create a new cart object passing it the request object 
+    product = get_object_or_404(Product, id=product_id) 
+    form = CartAddProductForm(request.POST)
+    if form.is_valid():
+        cd = form.cleaned_data
+        cart.add(product=product, quantity=cd['quantity'], update_quantity=cd['update'])
```
Comment (shivam04) [inline cart/views.py:15]: check if product is present or not

File: cart/views.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    cart = Cart(request)
+    product = get_object_or_404(Product, id=product_id)
+    cart.remove(product)
+    return redirect('cart:cart_detail')
+
+
+def cart_detail(request):
+    cart = Cart(request)
+    for item in cart:
+        item['update_quantity_form'] = CartAddProductForm(initial={'quantity': item['quantity'], 'update': True})
```
Comment (shivam04) [inline cart/views.py:29]: what is the need for it?

File: orders/models.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
+from django.db import models
+from shop.models import Product
+
+
+class Order(models.Model):
```
Comment (shivam04) [inline orders/models.py:5]: what if someone orders and wants the food to be delivered after some time?

File: orders/models.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    paid = models.BooleanField(default=False)
+
+    class Meta:
+        ordering = ('-created', )
+
+    def __str__(self):
+        return 'Order {}'.format(self.id)
+
+    def get_total_cost(self):
+        return sum(item.get_cost() for item in self.items.all())
```
Comment (shivam04) [inline orders/models.py:23]: check if item is number or not

File: orders/views.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+            order = form.save()
+            for item in cart:
+                OrderItem.objects.create(
+                    order=order,
+                    product=item['product'],
+                    price=item['price'],
+                    quantity=item['quantity']
+                )
+            cart.clear()
+        return render(request, 'orders/order/created.html', {'order': order})
```
Comment (shivam04) [inline orders/views.py:21]: instead of created.html redirect it to the homepage with a successful message

File: orders/views.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
+from django.shortcuts import render
+from .models import OrderItem
+from .forms import OrderCreateForm
+from cart.cart import Cart
+
+
+def order_create(request):
+    cart = Cart(request)
+    if request.method == 'POST':
+        form = OrderCreateForm(request.POST)
```
Comment (shivam04) [inline orders/views.py:10]: what if the form is  invalid how are you handling that case

File: cart/cart.py (hunk 1 above)
Comment (shadowshot-x) [reply cart/cart.py:1]: yes, it has been made especially for cart updation and deletion

File: cart/templates/cart/cart.html (hunk 2 above)
Comment (shadowshot-x) [reply cart/templates/cart/cart.html:4]: okk sir

File: cart/templates/cart/detail.html (hunk 3 above)
Comment (shadowshot-x) [reply cart/templates/cart/detail.html:26]: ok sir

File: cart/views.py (hunk 8 above)
Comment (shadowshot-x) [reply cart/views.py:29]: this is for updation of quantity of products

File: orders/views.py (hunk 11 above)
Comment (shadowshot-x) [reply orders/views.py:21]: Sir, The order number is generated over here. 
Should we add a javascript type message?

Comment (shivam04) [reply orders/views.py:21]: no use Django messages.

File: orders/models.py (hunk 10 above)
Comment (shadowshot-x) [reply orders/models.py:23]: sir, please elaborate

File: orders/views.py (hunk 12 above)
Comment (shadowshot-x) [reply orders/views.py:10]: okk sir, we will add a response for the same

File: orders/models.py (hunk 10 above)
Comment (shivam04) [reply orders/models.py:23]: ignore

File: cart/views.py (hunk 7 above)
Comment (shadowshot-x) [reply cart/views.py:15]: ok sir

File: orders/views.py (hunk 12 above)
Comment (shadowshot-x) [reply orders/views.py:10]: sir, if any field is invalid, The form will show a normal error and tell the user to correct it.

**Pull request closed without merging**